swarm/pss/client/client_test.go

```diff
@@ -0,0 +1,187 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/pot"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/pss"
+)
+
+func init() {
+	h := log.CallerFileHandler(log.StreamHandler(os.Stderr, log.TerminalFormat(true)))
+	log.Root().SetHandler(h)
+}
+
+func TestRunProtocol(t *testing.T) {
+	quitC := make(chan struct{})
+	ps := pss.NewTestPss(nil)
+	ping := &pss.PssPing{
+		QuitC: make(chan struct{}),
+	}
+	proto := newProtocol(ping)
+	_, err := baseTester(t, proto, ps, nil, nil, quitC)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	quitC <- struct{}{}
+}
+
+func TestIncoming(t *testing.T) {
+	quitC := make(chan struct{})
+	ps := pss.NewTestPss(nil)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	var addr []byte
+	ping := &pss.PssPing{
+		QuitC: make(chan struct{}),
+	}
+	proto := newProtocol(ping)
+	client, err := baseTester(t, proto, ps, ctx, cancel, quitC)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	client.ws.Call(&addr, "psstest_baseAddr")
+
+	code, _ := pss.PssPingProtocol.GetCode(&pss.PssPingMsg{})
+	rlpbundle, err := pss.NewProtocolMsg(code, &pss.PssPingMsg{
+		Created: time.Now(),
+	})
+	if err != nil {
+		t.Fatalf("couldn't make pssmsg")
+	}
+
+	pssenv := pss.PssEnvelope{
+		From:    addr,
+		Topic:   pss.NewTopic(proto.Name, int(proto.Version)),
+		TTL:     pss.DefaultTTL,
+		Payload: rlpbundle,
+	}
+	pssmsg := pss.PssMsg{
+		To:      addr,
+		Payload: &pssenv,
+	}
+
+	ps.Process(&pssmsg)
+
+	go func() {
+		<-ping.QuitC
+		client.cancel()
+	}()
+
+	select {
+	case <-client.ctx.Done():
+		t.Fatalf("outgoing timed out or canceled")
+	default:
```

Comment:
default casefalls through, the test can never fail, same for outgoing.
you need select here between ping.QuitC and ctx no?